### PR TITLE
Order last-n endpoints by video availability.

### PR DESCRIPTION
**Pull request by bgrozev**
When ordering the endpoints to forward, order not only by speech activity, but also whether endpoints are sending video (those sending video are put on the top of the list).

### CODE REVIEW

File: src/main/java/org/jitsi/videobridge/Conference.java [hunk 1]
```diff
@@ -1292,4 +1314,41 @@ public Object put(String key, Object value)
...
+        private void update(boolean force)
+        {
+            List<AbstractEndpoint> endpointsBySpeechActivity = speechActivity.getEndpoints();
+
+            Map<Boolean, List<AbstractEndpoint>> bySendingVideo
+                    = endpointsBySpeechActivity.stream()
+                        .collect(Collectors.groupingBy(AbstractEndpoint::isSendingVideo));
+
+            List<AbstractEndpoint> lastNEndpoints = new LinkedList<>();
+            lastNEndpoints.addAll(bySendingVideo.getOrDefault(true, Collections.emptyList()));
```
Comment (bbaldino) [inline src/main/java/org/jitsi/videobridge/Conference.java:1341]: I played with this code a bit and came up with a couple ideas.  Both are in Kotlin (I'd be all for changing `LastNEndpoints` to Kotlin, even if it meant it could no longer be a private inner class).

Option 1:
```
val (epsWithVideo, epsWithoutVideo) = endpointsBySpeechActivity.partition(AbstractEndpoint::isSendingVideo)
val lastNEndpointIds = (epsSending + epsNotSending).map(AbstractEndpoint::id)
```

Option 2: use a comparator
```
val lastNRankComparator = object : Comparator<Pair<Int, AbstractEndpoint>> {
    override fun compare(o1: Pair<Int, AbstractEndpoint>, o2: Pair<Int, AbstractEndpoint>): Int {
        // If both endpoints are sending/not sending video, compare by their original speaker rank
        if (o1.second.sendingVideo == o2.second.sendingVideo) {
            return o1.first.compareTo(o2.first)
        } else if (o1.second.sendingVideo) {
            return -1
        }
        return 1
    }
}
```
And then the code to sort:
```
val result = endpointsBySpeechActivity.mapIndexed { index, entry ->
    // Create an intermediary pair to store the original speaker rank
    Pair(index, entry)
}.sortedWith(lastNRankComparator)
```

File: src/main/java/org/jitsi/videobridge/Endpoint.java [hunk 2]
```diff
@@ -73,7 +73,7 @@
  */
 public class Endpoint
     extends AbstractEndpoint implements PotentialPacketHandler,
-        EncodingsManager.EncodingsUpdateListener
+        EncodingsManager.EncodingsUpdateListener, TransceiverEventHandler
```
Comment (bbaldino) [inline src/main/java/org/jitsi/videobridge/Endpoint.java:76]: For implementations of interfaces that will never be called by an external user of the class, I think it's nice to do "private" implementations of the interface (e.g. via a private inner class).  It also moves the methods out of the 'top level' of the class which I think is cleaner.

File: src/main/java/org/jitsi/videobridge/octo/OctoTransceiver.java [hunk 3]
```diff
@@ -54,18 +54,23 @@
     private final StreamInformationStore streamInformationStore
             = new StreamInformationStoreImpl();
 
-    private final OctoRtpReceiver octoReceiver;
+    final OctoRtpReceiver octoReceiver;
```
Comment (bbaldino) [inline src/main/java/org/jitsi/videobridge/octo/OctoTransceiver.java:57]: Can you keep this private and add a passthrough method in `OctoTransceiver`?

File: src/main/kotlin/org/jitsi/videobridge/octo/config/OctoRtpReceiver.kt [hunk 4]
```diff
@@ -60,6 +61,7 @@ import java.util.concurrent.atomic.AtomicBoolean
  */
 class OctoRtpReceiver(
     val streamInformationStore: ReadOnlyStreamInformationStore,
+    val eventHandler: TransceiverEventHandler?,
```
Comment (bbaldino) [inline src/main/kotlin/org/jitsi/videobridge/octo/config/OctoRtpReceiver.kt:64]: Why nullable?  It wasn't for the non-octo classes in JMT.

Comment (bbaldino) [reply src/main/kotlin/org/jitsi/videobridge/octo/config/OctoRtpReceiver.kt:64]: If there is a reason to keep it nullable, I'd add a default value of `null`.

Comment (bgrozev) [reply src/main/kotlin/org/jitsi/videobridge/octo/config/OctoRtpReceiver.kt:64]: The instance created in OctoEndpoints doesn't need an event handler

File: src/main/java/org/jitsi/videobridge/Conference.java (hunk 1 above)
Comment (bgrozev) [reply src/main/java/org/jitsi/videobridge/Conference.java:1348]: I agree Option 1 looks nicer (though that would be offset by having to pull it out of Conference), but I don't want to re-write and test it just because of that. Do you see a problem with this code?

Comment (bbaldino) [reply src/main/java/org/jitsi/videobridge/Conference.java:1348]: > Do you see a problem with this code?

Just that I think it's doing something simple but I find it hard to follow.  And I think being able to test it separately would be a good thing.

Comment (bgrozev) [reply src/main/java/org/jitsi/videobridge/Conference.java:1348]: I'll add tests in a separate PR, and I'll make sure to simplify the code. Can we move forward and ship this?

Comment (bbaldino) [reply src/main/java/org/jitsi/videobridge/Conference.java:1348]: Ok
